### PR TITLE
Pluggable driver

### PR DESCRIPTION
**Pull request by nathan-jenan-rancher**
This change gives rancher the ability to accept pluggable kontainer engine drivers
by adding the necessary types and removing the driver specific types.
    
Issue:
rancher/rancher#12698

### CODE REVIEW

File: apis/management.cattle.io/v3/cluster_types.go [hunk 1]
```diff
@@ -120,110 +122,31 @@ type ClusterCondition struct {
...
-	// Network
-	Network string `json:"network,omitempty"`
-	// Sub Network
-	SubNetwork string `json:"subNetwork,omitempty"`
-	// Configuration for LegacyAbac
-	EnableLegacyAbac            bool   `json:"enableLegacyAbac,omitempty"`
-	EnableStackdriverLogging    *bool  `json:"enableStackdriverLogging,omitempty" norman:"default=true"`
-	EnableStackdriverMonitoring *bool  `json:"enableStackdriverMonitoring,omitempty" norman:"default=true"`
-	MaintenanceWindow           string `json:"maintenanceWindow"`
+func init() {
```
Comment (cjellick) [inline apis/management.cattle.io/v3/cluster_types.go]: move to top of file

**Comment by StrongMonkey** [comment]
LGTM

File: apis/management.cattle.io/v3/schema/schema.go [hunk 2]
```diff
@@ -140,14 +140,6 @@ func clusterTypes(schemas *types.Schemas) *types.Schemas {
 		MustImport(&Version, v3.ImportClusterYamlInput{}).
 		MustImport(&Version, v3.ImportYamlOutput{}).
 		MustImport(&Version, v3.ExportOutput{}).
-		MustImportAndCustomize(&Version, v3.ETCDService{}, func(schema *types.Schema) {
```
Comment (StrongMonkey) [inline apis/management.cattle.io/v3/schema/schema.go]: hmm...did you accidentally remove this?

File: apis/management.cattle.io/v3/schema/schema.go [hunk 3]
```diff
@@ -253,7 +245,6 @@ func nodeTypes(schemas *types.Schemas) *types.Schemas {
 			schema.ResourceFields["clusterId"] = clusterField
 			schema.ResourceActions["cordon"] = types.Action{}
 			schema.ResourceActions["uncordon"] = types.Action{}
-			schema.ResourceActions["stopDrain"] = types.Action{}
```
Comment (StrongMonkey) [inline apis/management.cattle.io/v3/schema/schema.go]: is this related to your change?

File: apis/management.cattle.io/v3/kontainer_types.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	// Most recent observed status of the cluster. More info:
+	// https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#spec-and-status
+	Status KontainerDriverStatus `json:"status"`
+}
+
+type KontainerDriverStatus struct {
+	ActualURL      string      `json:"actualUrl"`
+	ExecutablePath string      `json:"executablePath"`
+	Conditions     []Condition `json:"conditions"`
+	DisplayName    string      `json:"displayName"`
```
Comment (StrongMonkey) [inline apis/management.cattle.io/v3/kontainer_types.go]: should displayName live in spec? i think that's how mapper works to find that field...

Comment (nathan-jenan-rancher) [reply apis/management.cattle.io/v3/kontainer_types.go]: It should not live in spec because it is a computed field, not settable by the user.  Also, the mapping is working correctly even with it in status instead of spec.

File: apis/management.cattle.io/v3/schema/schema.go (hunk 2 above)
Comment (nathan-jenan-rancher) [reply apis/management.cattle.io/v3/schema/schema.go]: Yes, I did.  Fixing now.

File: apis/management.cattle.io/v3/schema/schema.go [hunk 5]
```diff
@@ -578,3 +578,17 @@ func clusterCatalogTypes(schemas *types.Schemas) *types.Schemas {
 			}
 		})
 }
+
+func kontainerTypes(schemas *types.Schemas) *types.Schemas {
+	return schemas.
+		AddMapperForType(&Version, v3.KontainerDriver{},
```
Comment (cjellick) [inline apis/management.cattle.io/v3/schema/schema.go]: drop the generic config field

File: client/management/v3/zz_generated_kontainer_driver.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	KontainerDriverFieldUUID                 = "uuid"
+	KontainerDriverFieldWhitelistDomains     = "whitelistDomains"
+)
+
+type KontainerDriver struct {
+	types.Resource
+	Active               bool              `json:"active,omitempty" yaml:"active,omitempty"`
+	ActualURL            string            `json:"actualUrl,omitempty" yaml:"actualUrl,omitempty"`
+	Annotations          map[string]string `json:"annotations,omitempty" yaml:"annotations,omitempty"`
+	BuiltIn              bool              `json:"builtIn,omitempty" yaml:"builtIn,omitempty"`
```
Comment (cjellick) [inline client/management/v3/zz_generated_kontainer_driver.go]: make either readonly or drop it completely

File: apis/management.cattle.io/v3/schema_types.go [hunk 7]
```diff
@@ -51,6 +52,7 @@ type Field struct {
 	ValidChars   string   `json:"validChars,omitempty"`
 	InvalidChars string   `json:"invalidChars,omitempty"`
 	Description  string   `json:"description,omitempty"`
+	DynamicField bool     `json:"dynamicField,omitempty"`
```
Comment (ibuildthecloud) [inline apis/management.cattle.io/v3/schema_types.go]: What is DynamicField used for?

File: apis/management.cattle.io/v3/cluster_types.go [hunk 8]
```diff
@@ -61,19 +70,17 @@ type Cluster struct {
...
-	DockerRootDir                        string                               `json:"dockerRootDir,omitempty" norman:"default=/var/lib/docker"`
-	EnableNetworkPolicy                  *bool                                `json:"enableNetworkPolicy" norman:"default=false"`
+	DisplayName                          string                         `json:"displayName" norman:"required"`
+	Description                          string                         `json:"description"`
+	Internal                             bool                           `json:"internal" norman:"nocreate,noupdate"`
+	DesiredAgentImage                    string                         `json:"desiredAgentImage"`
+	ImportedConfig                       *ImportedConfig                `json:"importedConfig,omitempty" norman:"nocreate,noupdate"`
+	RancherKubernetesEngineConfig        *RancherKubernetesEngineConfig `json:"rancherKubernetesEngineConfig,omitempty"`
+	GenericEngineConfig                  *MapStringInterface            `json:"genericEngineConfig,omitempty"`
```
Comment (cjellick) [inline apis/management.cattle.io/v3/cluster_types.go]: add old types as map string interface. also, drop the fields from the schema. and add a controller that copies the data into generic field.

File: apis/management.cattle.io/v3/schema_types.go (hunk 7 above)
Comment (nathan-jenan-rancher) [reply apis/management.cattle.io/v3/schema_types.go]: When dynamic schemas are created or updated the schema created from the dynamic schema is merged with the existing schema.  The problem with this is that if a field is removed from the dynamic schema that field is not removed from the existing schema.  This makes it impossible to remove a dynamic field from a schema.

If dynamic field is set to true, then that field is removed from the existing schema before it is merged with the new schema.

The reason it has to exist, and isn't always just assumed to be true implicitly, is because the dynamic schema is actually marshaled to json, and then unmarshaled to the schema type. This only works because schema and dynamic schema are functionally identical despite being completely different types.  So while dynamic field will always be set to `True` on dynamic schema, it will be set to false on `Schema`.

**Comment by ibuildthecloud** [comment]
LGTM

File: apis/management.cattle.io/v3/schema/schema.go (hunk 3 above)
Comment (nathan-jenan-rancher) [reply apis/management.cattle.io/v3/schema/schema.go]: No, reverted this change.

File: apis/management.cattle.io/v3/schema/schema.go (hunk 5 above)
Comment (nathan-jenan-rancher) [reply apis/management.cattle.io/v3/schema/schema.go]: Done.

File: client/management/v3/zz_generated_kontainer_driver.go (hunk 6 above)
Comment (nathan-jenan-rancher) [reply client/management/v3/zz_generated_kontainer_driver.go]: Marked as `noupdate`.
Added logic to formatter.

File: apis/management.cattle.io/v3/cluster_types.go (hunk 8 above)
Comment (nathan-jenan-rancher) [reply apis/management.cattle.io/v3/cluster_types.go]: Done.
